### DISCUSSION TITLE
How to shutdown a well?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1462] [Q&A] [answered]
A well is included in a test case,
      <WellControls
        name="wellControls"
        type="injector"
        control="totalVolRate"
	useSurfaceConditions="1"
	surfacePressure="14432381.756"
	surfaceTemperature="368.15"
        referenceElevation="12.8961"
        targetBHP="1e8"
        targetTotalRate="0.004"
        injectionStream="{ 1.0, 0 }"/>

    <InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="mesh"
      polylineNodeCoords="{ { 5305, 5025, 15},
                            { 5305, 5025, 12.8961},
                            { 4695, 5025, 12.8961} }"
      polylineSegmentConn="{ { 0, 1 }, { 1, 2 } }"
      radius="0.09525"
      numElementsPerSegment="1">
      <Perforation
        name="injector1_perf1"
        distanceFromHead=".7155"/>
      <Perforation
        name="injector1_perf2"
        distanceFromHead="2.8194"/>
	  <Perforation
        name="injector1_perf3"
        distanceFromHead="200"/>
      <Perforation
        name="injector1_perf4"
        distanceFromHead="400"
		transmissibility="0"
	  />		
      <Perforation
        name="injector1_perf5"
        distanceFromHead="600"/>
      </InternalWell>

after 1 year, I would like to shut it down. How to shutdown this well in GEOSX?

Thanks.

**Comment by francoishamon** [comment]
Hello Deming,
Currently, we cannot shutdown a well easily, so you can only simulate the injection period. This is something that we also need to do, so we will implement it soon. Thanks for your patience

**Comment by francoishamon** [comment]
@wdmhouston Regarding your questions: _is it possible to shutdown a well? Is it possible to run a simulation without well? Is it possible to set 0 injection target rate to simulate the shutdown event._

Not possible to properly shutdown a well yet. 

Soon it will be possible to input the well rate as a table to make it time dependent, but there are some modifications to make in the well model to make sure that the problem is still well posed if the rate is zero. Currently, the zero-rate case is not going to work well.

For the type of simulation that you are running, the following workflow would make a lot of sense, even after we implement the capability to shut down a well:
1) Running the injection period with a well
2) Saving the values in each cell at the end of the injection period
3) Running the post-injection period without well, starting from the values saved in 2)

Ideally you would issue a restart file to do 2), and then you would use the restart file for 3) to start your new simulation. Presently this is not going to work because GEOSX expects the xml file to be the same in Steps 1) and 3).

So, instead of using the restart procedure in place in GEOSX, you can always get the values at the end of the injection period either from the HD5 file or the VTK file, and use a `TableFunction` for each primary variable to initialize your simulation in Step 3).

**Comment by wdmhouston** [comment]
Francois,

     Thanks for the clarification. I changed the injection rate to a very
small value in my restart case as a temp solution. Looking forward to the
new feature in the next release!


Regards,

Deming

On Wed, Jun 16, 2021 at 4:26 PM Francois Hamon ***@***.***>
wrote:

> @wdmhouston <https://github.com/wdmhouston> Regarding your questions: *is
> it possible to shutdown a well? Is it possible to run a simulation without
> well? Is it possible to set 0 injection target rate to simulate the
> shutdown event.*
>
> Not possible to properly shutdown a well yet.
>
> Soon it will be possible to input the well rate as a table to make it time
> dependent, but there are some modifications to make in the well model to
> make sure that the problem is still well posed if the rate is zero.
> Currently, the zero-rate case is not going to work well.
>
> For the type of simulation that you are running, the following workflow
> would make a lot of sense, even after we implement the capability to shut
> down a well:
>
>    1. Running the injection period with a well
>    2. Saving the values in each cell at the end of the injection period
>    3. Running the post-injection period without well, starting from the
>    values saved in 2)
>
> Ideally you would issue a restart file to do 2), and then you would use
> the restart file for 3) to start your new simulation. Presently this is not
> going to work because GEOSX expects the xml file to be the same in Steps 1)
> and 3).
>
> So, instead of using the restart procedure in place in GEOSX, you can
> always get the values at the end of the injection period either from the
> HD5 file or the VTK file, and use a TableFunction for each primary
> variable to initialize your simulation in Step 3).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GEOSX/GEOSX/discussions/1462#discussioncomment-880447>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQFWACEMKUZ3J76VUORALTTEJHHANCNFSM46RIW2AQ>
> .
>


-- 
Deming Wang
***@***.***

**Comment by francoishamon** [comment, marked as answer]
Hello @wdmhouston,
If you still need to shut down a well, this is now implemented properly. You will have to checkout branch `feature/tang/WellRateBHPTable`, recompile GEOSX, and follow this [example](https://github.com/GEOSX/GEOSX/blob/feature/tang/WellRateBHPTable/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml):

In the `CompositionalMultiphaseWell` block of the XML, you first use the keyword `targetTotalRateTableName` as below
https://github.com/GEOSX/GEOSX/blob/fddbc1fb2ed89b05c30c973f52e87e703940d5e7/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml#L49-L56
and then in the `Functions` block, you specify the rate as a function of time, with 0 if you want to shut down the well:
https://github.com/GEOSX/GEOSX/blob/fddbc1fb2ed89b05c30c973f52e87e703940d5e7/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml#L223-L227

**Reply by wdmhouston** [reply]
Thanks. will try!

**Reply by wdmhouston** [reply]
Francois,

    Do you have link of that example? That example link is not working any more, I guess because it's merged.

I'm asking for that example because I got the following message:
'''
Adding Output: TimeHistory, timeRateHistoryOutput4
   TableFunction: permxFunc
   TableFunction: permyFunc
   TableFunction: permzFunc
   TableFunction: poroFunc
   TableFunction: pressureFunc
   TableFunction: totalRateTable
***** ERROR
***** LOCATION: /apps/GEOSX/src/coreComponents/dataRepository/ObjectCatalog.hpp:190
***** Controlling expression (should be false): true
***** Rank 0: Could not find keyword "PoreVolumeCompressibleSolid" in this context. Please be sure that all your keywords are properly spelled or that input file parameters have not changed.
All available keys are: [
BiotPorosity,
BlackOilFluid,
BrooksCoreyBakerRelativePermeability,
BrooksCoreyCapillaryPressure,
BrooksCoreyRelativePermeability,
CarmanKozenyPermeability,
CO2BrineFluid,
CompositionalMultiphaseFluid,
CompressibleSinglePhaseFluid,
CompressibleSolidCarmanKozenyPermeability,
CompressibleSolidConstantPermeability,
CompressibleSolidParallelPlatesPermeability,
ConstantPermeability,
Contact,
Coulomb,
DamageElasticIsotropic,
DamageSpectralElasticIsotropic,
DamageVolDevElasticIsotropic,
DeadOilFluid,
DruckerPrager,
ElasticIsotropic,
ElasticOrthotropic,
ElasticTransverseIsotropic,
ExtendedDruckerPrager,
NullModel,
ParallelPlatesPermeability,
ParticleFluid,
PermeabilityBase,
PorousDruckerPrager,
PorousElasticIsotropic,
PorousElasticOrthotropic,
PorousElasticTransverseIsotropic,
PorousExtendedDruckerPrager,
PressurePorosity,
ProppantSlurryFluid,
StrainDependentPermeability,
TableRelativePermeability,
VanGenuchtenBakerRelativePermeability,
VanGenuchtenCapillaryPressure
]

...

I do have PoreVolumeCompressibleSolid keyword in my data file:

![image](https://user-images.githubusercontent.com/19333848/133525665-8459856a-f2b6-4405-ab96-97fe9d3f6dd7.png)

I also read that  the available keys list from log file, and it doesn't include PoreVolumeCompressibleSolid keysword.

any idea? maybe a bug?





> Hello @wdmhouston,
> If you still need to shut down a well, this is now implemented properly. You will have to checkout branch `feature/tang/WellRateBHPTable`, recompile GEOSX, and follow this [example](https://github.com/GEOSX/GEOSX/blob/feature/tang/WellRateBHPTable/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml):
> 
> In the `CompositionalMultiphaseWell` block of the XML, you first use the keyword `targetTotalRateTableName` as below
> https://github.com/GEOSX/GEOSX/blob/fddbc1fb2ed89b05c30c973f52e87e703940d5e7/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml#L49-L56
> 
> 
> and then in the `Functions` block, you specify the rate as a function of time, with 0 if you want to shut down the well:
> https://github.com/GEOSX/GEOSX/blob/fddbc1fb2ed89b05c30c973f52e87e703940d5e7/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml#L223-L227

**Reply by francoishamon** [reply]
Hi @wdmhouston, 
Yes you can try with this one for instance: https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml
It has a table that shuts down the well after some time.

We have changed a few things in the constitutive models recently, so if you want to update your input file, two solutions:

1- Post it and we will update it

or

2.a - Add `rockPorosity`, `nullSolid` in the `materialList` of the `CellElementRegion`
2.b - Add the following models in `Constitutive`:

```
<CompressibleSolidConstantPermeability
      name="rock"
      solidModelName="nullSolid"
      porosityModelName="rockPorosity"
      permeabilityModelName="rockPerm"/>

    <NullModel
      name="nullSolid"/>

    <PressurePorosity
      name="rockPorosity"
      defaultReferencePorosity="0.1"
      referencePressure="0.0"
      compressibility="1.0e-9"/>
```
with your values of `defaultReferencePorosity`, `referencePressure`, and `compressibility`, and remove `PoreVolumeCompressibleSolid`.
2.c - Remove the `FieldSpecification` block that enforces the `referencePorosity`

**Reply by wdmhouston** [reply]
> Hi @wdmhouston,
> Yes you can try with this one for instance: https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml
> It has a table that shuts down the well after some time.
> 
> We have changed a few things in the constitutive models recently, so if you want to update your input file, two solutions:
> 
> 1- Post it and we will update it
> 
> or
> 
> 2.a - Add `rockPorosity`, `nullSolid` in the `materialList` of the `CellElementRegion`
> 2.b - Add the following models in `Constitutive`:
> 
> ```
> <CompressibleSolidConstantPermeability
>       name="rock"
>       solidModelName="nullSolid"
>       porosityModelName="rockPorosity"
>       permeabilityModelName="rockPerm"/>
> 
>     <NullModel
>       name="nullSolid"/>
> 
>     <PressurePorosity
>       name="rockPorosity"
>       defaultReferencePorosity="0.1"
>       referencePressure="0.0"
>       compressibility="1.0e-9"/>
> ```
> 
> with your values of `defaultReferencePorosity`, `referencePressure`, and `compressibility`, and remove `PoreVolumeCompressibleSolid`.
> 2.c - Remove the `FieldSpecification` block that enforces the `referencePorosity`

Here is a test case, can you take a look? Thanks.
[wellcontroltest.zip](https://github.com/GEOSX/GEOSX/files/7177936/wellcontroltest.zip)

**Reply by francoishamon** [reply]
Hello @wdmhouston, 
Here is the updated file, let me know if you have issues.
[wellcontrolupdated.txt](https://github.com/GEOSX/GEOSX/files/7196520/wellcontrolupdated.txt)

**Comment by wdmhouston** [comment]
Thanks. I'll try.

On Wed, Sep 15, 2021 at 6:47 PM Francois Hamon ***@***.***>
wrote:

> Hi @wdmhouston <https://github.com/wdmhouston>,
> Yes you can try with this one for instance:
> https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml
> It has a table that shuts down the well after some time.
>
> We have changed a few things in the constitutive models recently, so if
> you want to update your input file, two solutions:
>
> 1- Post it and we will update it
>
> or
>
> 2.a - Add rockPorosity, nullSolid in the materialList of the
> CellElementRegion
> 2.b - Add the following models in Constitutive:
>
> <CompressibleSolidConstantPermeability
>       name="rock"
>       solidModelName="nullSolid"
>       porosityModelName="rockPorosity"
>       permeabilityModelName="rockPerm"/>
>
>     <NullModel
>       name="nullSolid"/>
>
>     <PressurePorosity
>       name="rockPorosity"
>       defaultReferencePorosity="0.1"
>       referencePressure="0.0"
>       compressibility="1.0e-9"/>
>
> with your values of defaultReferencePorosity, referencePressure, and
> compressibility, and remove PoreVolumeCompressibleSolid.
> 2.c - Remove the FieldSpecification block that enforces the
> referencePorosity
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GEOSX/GEOSX/discussions/1462#discussioncomment-1335929>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQFWDYG7OCV3CHDGTGAUDUCEWBTANCNFSM46RIW2AQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Deming Wang
***@***.***

**Reply by wdmhouston** [reply]
If I still want to manually specify porosity like before, how to do that?

here is what I did before, using FieldSpecification  and TableFunction .

```

<FieldSpecification fieldName="referencePorosity"
		                    functionName="poroFunc"
		                    initialCondition="1"
		                    name="referencePorosity"
		                    objectPath="ElementRegions/Reservoir"
		                    scale="1"
		                    setNames="{ all }"/>

<coordinateFiles="{xlin.geos,ylin.geos,zlin.geos}"
		               inputVarNames="{elementCenter}"
		               interpolation="nearest"
		               name="poroFunc"

```

**Reply by francoishamon** [reply]
```
<FieldSpecification
  fieldName="rockPorosity_referencePorosity"
  functionName="poroFunc"
  initialCondition="1"
  name="referencePorosity"
  objectPath="ElementRegions/Reservoir"
  scale="1"
  setNames="{ all }"/>
```
